src/com/amazon/ion/impl/lite/IonValueLite.java

```diff
@@ -614,7 +614,75 @@ public final SymbolToken getKnownFieldNameSymbol()
         return token;
     }
 
-    final boolean clearSymbolIDValues()
+    private static class ClearSymbolIDsHolder
+    {
+        boolean allSIDsClear = false;
+        IonContainerLite parent = null;
+        IonContainerLite.SequenceContentIterator iterator = null;
+    }
+
+    private static ClearSymbolIDsHolder[] growClearSymbolIDsHolderStack(
+            ClearSymbolIDsHolder[] clearSymbolIDsHolderStack)
+    {
+        ClearSymbolIDsHolder[] newClearSymbolIDsHolderStack = new ClearSymbolIDsHolder[clearSymbolIDsHolderStack.length * 2];
+        System.arraycopy(clearSymbolIDsHolderStack, 0, newClearSymbolIDsHolderStack, 0,
+                clearSymbolIDsHolderStack.length);
+        return newClearSymbolIDsHolderStack;
+    }
+
+    private boolean clearSymbolIDsIterative(boolean readOnlyMode) {
+        ClearSymbolIDsHolder[] clearSymbolIDsHolderStack = new ClearSymbolIDsHolder[CONTAINER_STACK_INITIAL_CAPACITY];
+        int clearSymbolIDsHolderStackIndex = -1;
+        IonValueLite value = this;
+        ClearSymbolIDsHolder currentClearSymbolIDsHolder = null;
+        boolean allSIDsClear = true;
+
+        do {
+            if (!(value instanceof IonContainerLite)) {
+                if (currentClearSymbolIDsHolder == null) {
+                    allSIDsClear = value.scalarClearSymbolIDValues();
+                } else {
+                    currentClearSymbolIDsHolder.allSIDsClear &= value.scalarClearSymbolIDValues();
+                }
+                if(readOnlyMode) {
+                    value._isLocked(true);
+                }
+            } else {
+                if (++clearSymbolIDsHolderStackIndex >= clearSymbolIDsHolderStack.length) {
+                    clearSymbolIDsHolderStack = growClearSymbolIDsHolderStack(clearSymbolIDsHolderStack);
+                }
+                currentClearSymbolIDsHolder = new ClearSymbolIDsHolder();
```

Comment:
We should do this only if the Holder at the new index is null. That way we do not have to allocate a new Holder each time we descend to the same depth. You'll need to delete line 677 below, but you could replace it with a line that clears out the holder's iterator so that it may be garbage collected.